File: leetcode/moveZeroes/tom/moveZeroes.py
"""
Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Example:

Input: [0,1,0,3,12]
Output: [1,3,12,0,0]
Note:

You must do this in-place without making a copy of the array.
Minimize the total number of operations.


"""

class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: None Do not return anything, modify nums in-place instead.
        """
        pt1 = 0      
        pt2 = 0

        def swap(pt1,pt2):
            tmp = nums[pt1]
            nums[pt1] = nums[pt2]
            nums[pt2] = tmp

        while pt2 < len(nums) :      ### pt1 will not increment when it encounters zero.
            if nums[pt2] != 0 :    
                if pt2 == pt1 :
                    pt2 += 1         #if pt2 == pt1, it means pt1 doesnt meet zero yet. therefore keep moving forward
                    pt1 += 1
                else :
                    swap(pt1,pt2)    # if pt2 != pt1, swap them, and then moving pt1 to the next zero value.
                    pt2 += 1
                    while nums[pt1] != 0 :
                        pt1 += 1
            else :             ### pt1 will not increment when encountering zero. pt2 will keep incrementing until it meet non-zero value
                pt2 += 1
        return nums

File: leetcode/moveZeroes/guanduo/moveZeroes.cpp
/*
Given an array nums, write a function to move all 0's to the end of it while maintaining the relative order of the non-zero elements.

Example:

Input: [0,1,0,3,12]
Output: [1,3,12,0,0]
Note:

You must do this in-place without making a copy of the array.
Minimize the total number of operations.
*/

#include <iostream>
#include <vector>
/*
modify the array in place.
Iterate through the array. If equal to zero, skip incrementing the tracking index.
After the entire iteration, assign 0 to all the indices that is covered by the difference between tracking index and size.
 */

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
      int track = 0;//this guy will increment if not 0. Will skip incrementing if 0

      for(int i=0;i<nums.size();i++)
	if (nums[i] != 0)  nums[track++] = nums[i]; //increment track if non-zero

      while(track<nums.size())//add all the zeroes
	nums[track++] = 0;
    }
};
